Yet more vision fixes

phil235 opened the pull request:
- Fixes waking up immediately every life() call when having reagents giving you sleeping += 1. Fixed by changing the values a bit but the effect given is the same. Fixes #15322
- Fixes similar issue when you have one reagent making you asleep or weakened, and another waking you up.

phil235 committed:
Fixes waking up immediately every life() call when having reagents giving you sleeping += 1. Fixed by changing the values a bit but the effect given is the same.
Fixes revenant not seeing itself and not having his action buttons.
Fixes many mobs not having an hud_used.

phil235 committed:
Fixes fullscreen overlay disappearing when toggling the hud.
Fixes our vision when being ejected by a delivery chute and when inside a disposal unit that is being destroyed.

neersighted commented on code/modules/recycling/disposal-unit.dm:
```diff
@@ -194,7 +194,7 @@
 /obj/machinery/disposal/proc/eject()
 	for(var/atom/movable/AM in src)
-		AM.loc = src.loc
+		AM.forceMove(src.loc)
```
Any reason not to use get_turf?

phil235 replied:
I actually hesitated to use get_turf(), I can change it if you want.

neersighted replied:
I'd consider it more idomatic

neersighted commented on code/modules/recycling/disposal-unit.dm:
```diff
@@ -251,7 +251,7 @@
 			target = get_offset_target_turf(src.loc, rand(5)-rand(5), rand(5)-rand(5))
 
-			AM.loc = src.loc
+			AM.forceMove(src.loc)
```
Ditto

phil235 committed:
I removed all immediate health/status/stamina/canmove updates from the reagents' on_mob_life(). The potential updates now only happens once at the end of the metabolization process. So that we do not call updates after every effect of every reagent, which causes issue if we have one reagent paralysing you and another removing your paralysis (you end up just passing out for a split second every life() call ).
Each reagent on_mob_life() and overdose and addiction procs now returns 1 if it requires us to call mob updates (so that we do not call these updates for reagents that do not affect your status/heath/stamina/mobility).
I added an argument to more adjustments proc (adjustSleeping, adjustBruteLoss, etc) to not automatically call the relevant mob update.

phil235 commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -270,24 +270,24 @@ var/const/INJECT = 5 //injection
+	if(M && need_mob_update)
+		M.updatehealth()
+		M.update_canmove()
+		M.update_stamina()
```
If one of the metabolized reagent changed our mob's health or stat or mobility or stamina, we update all these here and only once, instead of multiple times inside each reagent's on_mob_life().

I didn't think necessary to separate the possible updates (i.e. if(updatehealth_needed) , if(update_stamina_needed), etc). Having on_mob_life() return a list of all the type of updates needed seemed a bit too much.

phil235 commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -361,30 +363,33 @@
+				else
+					H.faction -= "slime"
+			else
+				H << "<span class='danger'>The pain vanishes suddenly. You feel no different.</span>"
```
this fixes the sleep(30) putting life() on hold for 3 seconds.

phil235 commented on code/modules/mob/living/damage_procs.dm:
```diff
@@ -73,7 +73,6 @@
 		if(JITTER)
 			if(status_flags & CANSTUN)
 				jitteriness = max(jitteriness,(effect * blocked))
-	updatehealth()
```
not needed. If an update is required, Weaken(), Paralyse() or Stun() will call it.

PKPenguin321 commented:
please please please fix this and merge it

phil235 commented:
@PKPenguin321 I'm waiting for my other PRs to be merged first so I don't have to fix merge conflicts multiple times.

tkdrg commented:
@phil235 i think it should be fine to fix this one now maybe? ping me once conflicts are fixed

tkdrg commented:
@phil235 How's the status on this?
